applets/notification_area/na-grid.c

```diff
@@ -0,0 +1,523 @@
+/*
+ * Copyright (C) 2002 Red Hat, Inc.
+ * Copyright (C) 2003-2006 Vincent Untz
+ * Copyright (C) 2007 Christian Persch
+ * Copyright (C) 2017 Colomban Wendling <cwendling@hypra.fr>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/* Well, actuall'y is the Tray itself, the container for the items.  But
+ * NaTray is already taken for the XEMBED part, so for now it's called NaGrid,
+ * don't make a big deal out of it. */
+
+#include "config.h"
+#include <gtk/gtk.h>
+
+#include "na-grid.h"
+
+#include "system-tray/na-tray.h"
+#include "status-notifier/sn-host-v0.h"
+
+#define MIN_ICON_SIZE_DEFAULT 26
+#define USE_ONLY_ONE_LINE_DEFAULT TRUE
+
+typedef struct
+{
+	gint cols;
+	gint rows;
+	gint length;
+} GridProps;
+
+typedef struct
+{
+	GtkOrientation orientation;
+	guint          index;
+	NaGrid        *na_grid;
+} PackData;
+
+struct _NaGrid
+{
+	GtkGrid    parent;
+
+	gint       icon_padding;
+	gint       icon_size;
+	gint       min_icon_size;
+	gboolean   use_only_one_line;
+
+	GSList    *hosts;
+	GSList    *items;
+	GridProps *grid_props;
+};
+
+enum
+{
+	PROP_0,
+	PROP_ICON_PADDING,
+	PROP_ICON_SIZE
+};
+
+G_DEFINE_TYPE (NaGrid, na_grid, GTK_TYPE_GRID)
+
+static gint
+compare_items (gconstpointer a,
+               gconstpointer b)
+{
+	NaItem *item1;
+	NaItem *item2;
+	NaItemCategory c1;
+	NaItemCategory c2;
+	const gchar *id1;
+	const gchar *id2;
+
+	item1 = (NaItem *) a;
+	item2 = (NaItem *) b;
+
+	c1 = na_item_get_category (item1);
+	c2 = na_item_get_category (item2);
+
+	if (c1 < c2)
+		return -1;
+	else if (c1 > c2)
+		return 1;
+
+	id1 = na_item_get_id (item1);
+	id2 = na_item_get_id (item2);
+
+	return g_strcmp0 (id1, id2);
```

Comment:
If you need to do a comparison here, I think you must to do:

`return (g_strcmp0 (id1, id2) == 0);`

Without `"== 0"` , it doesn't work as expected.